qiskit/circuit/quantumcircuit.py

```diff
@@ -312,6 +318,12 @@ def from_instructions(
         circuit = QuantumCircuit(name=name, global_phase=global_phase, metadata=metadata)
         added_qubits = set()
         added_clbits = set()
+        if qubits:
+            circuit.add_bits(qubits)
+            added_qubits.update(qubits)
+        if clbits:
+            circuit.add_bits(clbits)
+            added_clbits.update(clbits)
```

Comment:
Oh actually a second thing sorry - since you type hinted these as `Iterable`, we need to consume them into a reusable sequence type in order for the Boolean conversion and double usage to work correctly with things like generators.

Reply:
Good catch. I think the boolean conversion is fine though, right? I also added a test for this case.

Reply:
The magic Boolean conversion for sequence-likes depends on having a `__len__` (then the conversion is `__len__() != 0`) - generators don't have that, so they fall back to the `object.__bool__` which is just universally `True`:
```python
>>> bool((x for x in ()))
True
```
If we just move the `list(qubits)` call up above the `if qubits` check it'll work fine.  If you make it `tuple`, then the default case is a no-op (`a = (); tuple(a) is a` is `True`), but there's no chance that would ever be a performance concern anyway.

Reply:
I understand that generators are truthy, but that seems to imply that what I wrote is correct because we want the code block in the if statement to execute for generators. If what I wrote is incorrect, could you provide a counterexample?

Reply:
Oh yeah, I get what you mean, sorry.  Yeah, you're right, it's fine this way.